<a href="https://colab.research.google.com/github/dyjdlopez/deep-learning-fundamentals/blob/main/module-1/mldl_leclab_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lecture 2: Fundamentals of Linear Algebra for Machine Learning and Data Science (Part 1)
### Deep Learning Course 2023
$_\text{Engr. D.J.D Lopez, M.Sc.}$<br>
$_\text{Adjunct Professor, Adamson University}$

## Data Representation

Data and information can be stored electronically using digital data. Digital data in its true form is considered quasi-structured data due to its nature-specific interpretability. To make it easier for analysis, data can be re-roganized as structured data.

## Linear Algebra
A field of mathematics that deals with vector analysis and their transformations.

The discussion of linear algebra revolves around vectors, their representations, interpretation, and transformation.

# Linear Algebra and Machine Learning

Machine learning deals with the analysis of information, linear algebra can be used to model the algorithms and transformation of data to get their **essence** or patterns.

## Let's start coding Linear Algebra
The idea of doing linear algebra in code is part of computational programming. We take advantage of the logic and arithmetic of linear algebra and code them to apply the theory. We can immediately see the transformation of data and how they are represented.

In this discussion, we will be using `numpy` as the main library for computation and `matplotlib` for visualizing the data.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(200)

## Vectors
Vectors are the most fundamental concept in linear algebra. Vectors in the computer science aspect are the collection of **scalars** or numbers; or in the mathematical perspective, these are a representation of a point in space. We can mathematically represent vectors as:
$$\mathbf{v} = \begin{pmatrix} a_0 & a_1 & a_2 & ... & a_{n-1} & a_n \end{pmatrix} $$

In [ ]:
## Creating a Scalar
scalar = 1
## in numpy
scalar_np = np.array(2)

print(scalar)
print(scalar_np)

1
2


In [ ]:
## Creating a vector in numpy
vector = np.array([1,2,3])

print(vector)

[1 2 3]


## Matrices
Matrices are a collection of scalars that could be represented in a lattice-like arrangement with $m$ rows and $n$ columns. It takes on a form like this:
$$\mathbf{M} = \begin{pmatrix} a_{00} & a_{01} & a_{02} & ...  & a_{0n} \\ a_{10} & a_{11} & a_{12} & ... & a_{1n} \\ a_{20} & a_{21} & a_{22} & ... & a_{2n} \\ \vdots & \vdots & \vdots & \ddots & \vdots\\ a_{m0} & a_{m1} & a_{m2} & ... & a_{mn} \end{pmatrix} $$


In [ ]:
## Creating a Matrix (elements)
A = np.array([
    [1,2,3],
    [3,1,0],
    [-1,0,2]
])
A

array([[ 1,  2,  3],
       [ 3,  1,  0],
       [-1,  0,  2]])

You can also think that matrices are collections of scaled vectors like this:
$$\mathbf{M} = \begin{pmatrix} a_0 \cdot v_0 \\ a_1 \cdot v_1 \\ \vdots \\ a_n\cdot v_n \end{pmatrix}$$

In [ ]:
## Creating a Matrix (vectors)
v1 = np.array([1,2,3])
v2 = np.array([3,2,1])
v3 = np.array([1,-1,-2])
B = np.array([
    2*v1,
    -1*v2,
    0.5*v3
])
B

array([[ 2. ,  4. ,  6. ],
       [-3. , -2. , -1. ],
       [ 0.5, -0.5, -1. ]])

# Vectorization
The idea of vectorization is coding with linear algebra in mind, this includes the representation of data and the operations that we can do on vectors such as element-wise operations, broadcasting, inner products, and even projections.

# Matrix Operations
Before we get to fully vectorize code, we need to refresh in performing in operating with vectors. In the following slides, we'll review the arithmetics we can do with vectors, transposition, the inner product, vector norms, projection, and solving linear equations.

## Element-wise Arithmetic
The element-wise and scalar arithmetic can be done in Python such as addition, subtraction, multiplication, division, exponentiation, and even other transcendental transformations.

In [ ]:
### Element-wise operations
A = np.array([
    [1,2],
    [-1,0]
])
B = np.array([
    [0,1],
    [1,0]
])

In [ ]:
### Scalar operations
k = -1
A = np.array([
    [1,-1],
    [2,3]
])

In [ ]:
### Element-wise vector transformations
f = lambda x: np.exp(x)
g = lambda x: np.log(x)
A = 5*np.ones((2,2))

## Transposition
The transposition operation flips the matrix along its diagonal. In mathematical terms it will for some matrix $\mathbf{M}$ its transposed form will be $\mathbf{M}^T$:
$$\mathbf{M} = \begin{pmatrix} a_{00} & a_{01} & a_{02} \\ a_{10} & a_{11} & a_{12} \\ a_{20} & a_{21} & a_{22}\end{pmatrix} ; \mathbf{M}^T = \begin{pmatrix} a_{00} & a_{10} & a_{20} \\ a_{01} & a_{11} & a_{21} \\ a_{02} & a_{12} & a_{22}\end{pmatrix}$$


In [ ]:
## Transposing a square matrix
M = np.random.randint(1,5,(3,3))
np.transpose(M)

array([[3, 1, 1],
       [2, 3, 4],
       [1, 4, 2]])

In [ ]:
## Transposing non-square matrices
N = np.random.randint(2,6,(4,3))
N.T

array([[4, 3, 4, 5],
       [5, 5, 2, 5],
       [3, 4, 3, 5]])

In [ ]:
## Transposing row/column vectors
v = np.random.randint(2,6,(1,3))
v.T

array([[3],
       [5],
       [3]])

### Inner Product
The inner product is one of the most important operations in linear algebra. It solves for the linear combination of the vectors and matrices. The inner product is not always commutative. 

The inner product of a matrix can be solved only if the number of columns of the first vector is equal to the number of rows in the second vector. So if the first matrix is $\mathbf{A}$ has a shape of $(m,n)$ and and the second matrix is $\mathbf{B}$ has a shape of $(i,j)$, then you can perform $AB$ if $n=i$.

In [ ]:
## Row-column inner product
A = np.random.randint(1,10,(1,3))
B = np.random.randint(1,10,(1,3))

In [ ]:
## Non-square inner product
A = np.random.randint(1,10,(2,3))
B = np.random.randint(1,10,(3,2))

In [ ]:
## Square inner product
A = np.random.randint(1,10,(2,2))
B = np.random.randint(1,10,(2,2))

## The Vector Norm
The norm of a vector describes its magnitude along a dimension. In calculations in machine learning we often use norms in the first and second dimensions, we call these the L1 and L2 norm. The general formula for the vector norm is called the Frobenius norm:
$$||\mathbf{M}||_n = \Biggl(\sum^N_{i=0}m_i^n\Biggr)^{1/n}$$

In [ ]:
## Let's translate the Frobenius norm as code:
def frobenius_norm(n, vector):
    n_sum = np.sum(vector**n)
    vect_norm = np.power(n_sum,1/n)
    return vect_norm

## The L1 Norm
The L1 norm is also called the Manhattan distance. Using the Frobenius norm, we set the value of $n$ to 1 signifying the dimensionality as 1. So we get:
$$||\mathbf{M}||_1 = \sum^N_{i=0}m_i$$

In [ ]:
def l1_norm(vector):
    vect_norm = np.sum(vector)
    return vect_norm

In [ ]:
v = np.array([1,1,2])

## using our l1 function
# l1_norm(v)

## using our frobenius norm function
# frobenius_norm(1, v)

## The L2 Norm
The L2 norm or the Euclidean norm, shows the vector magnitude in the 2D plane. Using the Frobenius norm, we set the value of $n$ to 2 signifying the dimensionality as 2. So we get:
$$||\mathbf{M}||_n = \Biggl(\sum^N_{i=0}m_i\Biggr)^{1/2} = \sqrt{\sum^N_{i=0}m_i}$$

In [ ]:
def l2_norm(vector):
    n_sum = np.dot(vector, vector)
    vect_norm = np.sqrt(n_sum)
    return vect_norm

In [ ]:
v = np.array([1,1,2])
frobenius_norm(1, v)

4.0

Numpy also provides a library for the Frobenius norm. 
You can check the docs [here](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html)

In [ ]:
np.linalg.norm(v,ord=1)

4.0

## Normalized Vectors
In most cases, we want matrices to hold unit vectors. Meaning, we would have the values of our vectors only between 0 and 1. To do this we can perform normalization to our vector or matrix $\mathbf{M}$:
$$\mathbf{M}_{norm} = \frac{\mathbf{M}}{||M||}$$

In [ ]:
def vector_norm(vector):
    return vector/np.linalg.norm(vector)

In [ ]:
v = np.array([50,32, 13])
vector_norm(v)

array([0.8227736 , 0.52657511, 0.21392114])

# Checkpoint Activity 1

## Cosine Similarity
We can determine the similarity of vectors by measuring the angle between them. We can do this using the cosine similarity theorem. Consider two vectors $A$ and $B$:
$$\cos{\theta} = \frac{AB}{||A||\cdot||B||}$$

### Problem Case 1
In Natural Language Processing, words can be represented by vectors in a process called "word embeddings" and vectorization. Supposed we were given a corpus (a piece of text) and we are tasked to determine the similarity of the words **conflagration, mayhem, and cataclysm**. Given in the code below are the vector embeddings of the words, determine the similarities of each word to each other.

In [ ]:
w1 = np.array([0.0513, 0.9788]) ##conflagration
w2 = np.array([0.7540, 0.8824]) ##mayhem
w3 = np.array([0.6115, 0.9788]) ##cataclysm

In [ ]:
def cos_similarity(a,b):
    inner_prod = a@b
    norm_a = np.linalg.norm(a)
    norm_b = np.linalg.norm(b)
    similarity = inner_prod / (norm_a*norm_b)
    return similarity    

In [ ]:
## Brute force
words = [w1,w2,w3]
for i in range(len(words)):
    for j in range(len(words)):
        cossim = cos_similarity(words[i], words[j])
        print(f'Similarity between w{i} and w{j}: {cossim}')

An important property of matrices is their **orthogonality**. Like in geometry, this signifies that the vectors form 90 degrees with each other. Using the idea of cosine similarity, vectors are similar id $\cos(\theta) = 1$ which indicates $\theta = 90$ or $\frac{\pi}{2}$. Another property called **orthonormality** tells us that the matrix is orthogonal and normal at the same time.

## Matrix Inverse
The inverse of a matrix is similar to the concept of inverse functions. In a sense, matrices are linear transformations that are a form of function. The inverse of a function can be solved in numerous ways using determinants or co-factors. Since that is already covered in either College-level algebra or Linear Algebra courses, we will skip ahead with the direct implementation. We can then further denote the inverse of a matrix $\mathbf{M}$ as:
$$\mathbf{M}^{-1}$$
With the notion that the matrix inverse is simlilar inverse functions, we can algebraically cancel out matrices in an equation by doing a dot product between a matrix and itself:
$$\mathbf{M}^{-1}\cdot \mathbf{M} = I$$

We can perform matrix inversion in `numpy` using the following codes:

In [ ]:
M = np.array([
    [1,2],
    [-1,0]
])
M_inv = np.linalg.inv(M)
M_inv

array([[ 0. , -1. ],
       [ 0.5,  0.5]])

In [ ]:
M_inv @ M

array([[1., 0.],
       [0., 1.]])

Take note that matrix inversion specifically with `np.linalg.inv()` assumes that the matrix is square. An error will occur when you input a non-square matrix in the function. To resolve this, we turn to the concept of the pseudo-inverse. We will discuss the pseudo-inverse operation further in the later lessons but the main idea is transforming the input matrix $\mathbf{M}$ into a [positive semi-definite matrix](https://www.youtube.com/watch?v=xsP-S7yKaRA) before performing an inversion. In `numpy`, we can use the following code:

In [ ]:
M = np.array([
    [1,2],
    [-1,0],
    [1,1]
]) ## a non-square matrix
M_pinv = np.linalg.pinv(M) ## pseudo-inverse
M_pinv

array([[-1.66666667e-01, -8.33333333e-01,  3.33333333e-01],
       [ 5.00000000e-01,  5.00000000e-01, -5.64008002e-17]])

# Checkpoint Activity 2

## System of Linear Equations
Solving a system of linear equations is an essential computational technique in engineering. The **system** part of the term refers to a collection of related instances or behavior of entities. These behaviors could be formulated or characterized as **linear equations**. Since this kind of formulation is also a collection of linear transformations, we can also represent it in matrix form.

### Problem Case 2

You are a coffee shop owner. For a week you have a quota to sell 650 cups of coffee wherein you have a pre-order of 58 cups from the city hall every Monday. The rest of the week you notice that the average consumption is 100 cups per day. Every week you start with supplies that are approximately 2,800 cups of coffee. You notice your shop produces 200 cups per day.  Will you meet your quota? If so, what day of the week will you meet your quota? If not, what adjustments will you make to meet your quota?

This is a supply and demand problem. This problem requires you to determine the individual equations for the supply and demand of coffee. 

Let's denote the demand equation as $D$.
$$D(t)=100t+58$$
Where $t$ is time in days and 58 is our starting point every week (Mondays). We now need to determine $t$ to satisfy the equation:
$$D(t) = 650$$
We can reformulate the equation as:
$$100t + 58 = 650$$

Now for the supply equation. We can denote it as $S$.
$$S(t)=-200t+2,800$$
Where $t$ is time in days and 2,800 is our initial supply for the week. The rate of consumption is 200 cups per day, we make it negative since our supply should diminish over time. we assume that at the end of the week we will empty out our supplies.
$$S(t) = 0$$
We can finally reformulate it as:
$$-200t+2,800 = 0$$

Now since they are part of a single system we can rewrite the equations as:

$$
\left\{
    \begin{array}\\
        100t+58=650\\ 
        -200t + 2,800=0
    \end{array}
\right.$$

Or using the augmented matrix form:
$$\begin{bmatrix}100&58\\-200&2,800\end{bmatrix} \cdot \begin{bmatrix}t\\b\end{bmatrix}=\begin{bmatrix}650\\0\end{bmatrix}$$

To solve for the values of $t$ (day of the week) and $b$ (ratio of much coffee we should have for the week) we can solve this using linear algebra using the inverse function:
$$ \begin{bmatrix}t\\b\end{bmatrix}=\begin{bmatrix}100&58\\-200&2,800\end{bmatrix}^{-1} \cdot\begin{bmatrix}650\\0\end{bmatrix}$$

Let's convert the formulation in code:

In [ ]:
D =  np.array([100, 58])
S = np.array([-200, 2800])
y = np.array([650, 0])
X = np.array([D,S])
np.linalg.inv(X)@y

array([6.24142661, 0.44581619])

According to our code the resulting vector is $\begin{bmatrix}6.24&0.44\end{bmatrix}^T$ which corresponds to our variable array $\begin{bmatrix}t&b\end{bmatrix}^T$. Since $t=6.24$ we round it up to the nearest ones so $t=7$, this means that we will meet our quota on Sunday.

# Module Checkpoint

### The Gramian Matrix
A special type of inner product is the inner product between the transpose of a matrix and itself. We can mathematically represent the [Gramian matrix](https://en.wikipedia.org/wiki/Gram_matrix) as:
$$G = X^T\cdot X$$
The Gramian/ gram matrix is an important matrix in machine learning. It can be used in several applications such as analysis of parameters in a dataset, a data representation of a dataset, and as a regularization term in regression algorithms.

As a data scientist, you are given the task to do data exploration. You were given a very small dataset resulting from an AB Test for a new formulation of boba milk tea. The data collected only has 5 respondents and two factors were considered: boba texture, tea quality, and sweetness. Instead of integer values, these parameters were converted to continuous float values. You were wierded out but proceeded with the analysis anyways. You wanted to know the level of relatedness of each factors.

Let's say we split the test data for the A and B test, we would have two $5\times 3$ matrices.
$$A = \begin{bmatrix}
0.67&0.56&0.21\\
0.78&0.43&0.18\\
0.44&0.28&0.02\\
0.88&0.88&0.88 \\
0.88&0.87&0.83
\end{bmatrix}; 
B = \begin{bmatrix}
0.89&0.77&0.60\\
0.90&0.87&0.61\\
0.74&0.90&0.70\\
1.0&0.91&0.90\\
0.88&0.88&0.88
\end{bmatrix}$$

In [ ]:
A = np.array([
    [0.67, 0.56, 0.21],
    [0.78, 0.43, 0.18],
    [0.55, 0.32, 0.11],
    [0.88, 0.88, 0.88],
    [0.88, 0.87, 0.83]
])
B = np.array([
    [0.89, 0.77, 0.60],
    [0.90, 0.87, 0.61],
    [0.74, 0.90, 0.70],
    [1.0, 0.91, 0.90],
    [0.88, 0.88, 0.88]
])

In [ ]:
## Compute for the Gramian Matrix of A
G_a = A.T@A
G_a

array([[2.9086, 2.4266, 1.8464],
       [2.4266, 2.1322, 1.7267],
       [1.8464, 1.7267, 1.5519]])

Now you might say that they don't like correlations since the diagonal of the matrix are not high even though they are supposed to be correlations of themselves. To make it similar to the cosine similarity, we would need to divide the matrix with the product of the norms of the features. We can do the following adjustments:

In [ ]:
A_norms = np.linalg.norm(A, axis=0, keepdims=True)
G_a/(A_norms.T @ A_norms)

array([[1.        , 0.9744111 , 0.86906433],
       [0.9744111 , 1.        , 0.94922912],
       [0.86906433, 0.94922912, 1.        ]])